tests/learning/test_discovery_phases.py

```diff
@@ -0,0 +1,105 @@
+"""
+This file is part of nucypher.
+
+nucypher is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+nucypher is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+from unittest.mock import patch
+
+import maya
+import pytest
+import time
+from flask import Response
+from umbral.keys import UmbralPublicKey
+
+from nucypher.characters.lawful import Ursula
+from tests.performance_mocks import mock_cert_storage, mock_cert_loading, mock_verify_node, \
+    mock_message_verification, \
+    mock_metadata_validation, mock_signature_bytes, mock_stamp_call, mock_pubkey_from_bytes, VerificationTracker, \
+    NotARestApp, mock_secret_source, NotAPublicKey
+
+"""
+Node Discovery happens in phases.  The first step is for a network actor to learn about the mere existence of a Node.
+This is a straightforward step which we currently do with our own logic, but which may someday be replaced by something
+like libp2p, depending on the course of development of those sorts of tools.  The introduction of hamming distance
+in particular is useful when wanting to learn about a small number (~500) of nodes among a much larger (25,000+) swarm.
+This toolchain is not built for that scenario at this time, although it is not a stated nongoal. 
+
+After this, our "Learning Loop" does four other things in sequence which are not part of the offering of node discovery tooling alone:
+
+* Instantiation of an actual Node object (currently, an Ursula object) from node metadata.
+* Validation of the node's metadata (non-interactive; shows that the Node's public material is indeed signed by the wallet holder of its Staker).
+* Verification of the Node itself (interactive; shows that the REST server operating at the Node's interface matches the node's metadata).
+* Verification of the Stake (reads the blockchain; shows that the Node is sponsored by a Staker with sufficient Stake to support a Policy).
+
+These tests show that each phase of this process is done correctly, and in some cases, with attention to specific
+performance bottlenecks.
+"""
```

Comment:
This utterly brilliant piece of commentary belongs to a higher-order artifact (nucypher docs?), rather than just to a test header.

Reply:
I'm talking about all previous paragraphs, not just the immediately above (which in fact is only relevant to the test)

Reply:
Oh, uhh - thanks!  Where does it belong then?

Reply:
I'm not sure. Perhaps we can just open an issue/TODO about finding a place to put this kind of development documentation which is not immediately useful for users, but it's gold for us.